telemetry/telemetry.go

```diff
@@ -0,0 +1,259 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/user"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/akitasoftware/akita-cli/cfg"
+	"github.com/akitasoftware/akita-cli/printer"
+	"github.com/akitasoftware/akita-cli/rest"
+	"github.com/akitasoftware/akita-cli/version"
+	"github.com/akitasoftware/akita-libs/analytics"
+)
+
+var (
+	// Shared client object
+	analyticsClient analytics.Client
+
+	// Client key; set at link-time with -X flag
+	defaultSegmentKey = ""
+
+	// Store the distinct ID; run through the process
+	// of getting it only once.
+	userDistinctID   string
+	userDistinctOnce sync.Once
+)
+
+type nullClient struct{}
+
+func (_ nullClient) TrackEvent(_ *analytics.Event) error {
+	return nil
+}
+
+func (_ nullClient) Track(distinctID string, name string, properties map[string]any) error {
+	return nil
+}
+
+func (_ nullClient) Close() error {
+	return nil
+}
+
+func init() {
+	// Opt-out mechanism
+	disableTelemetry := os.Getenv("AKITA_DISABLE_TELEMETRY")
+	if disableTelemetry != "" {
+		if val, err := strconv.ParseBool(disableTelemetry); err != nil && val {
+			printer.Stderr.Infof("Telemetry disabled.\n")
+			analyticsClient = nullClient{}
+			return
+		}
+	}
+
+	// If unset, will be "" and we'll use the default
+	segmentEndpoint := os.Getenv("AKITA_SEGMENT_ENDPOINT")
+
+	// If unset, will use this hard-coded value.
+	segmentKey := os.Getenv("AKITA_SEGMENT_WRITE_KEY")
+	if segmentKey == "" {
+		segmentKey = defaultSegmentKey
+	}
+	if segmentKey == "" {
+		printer.Stderr.Warningf("Telemetry unavailable; no Segment key configured.\n")
+		analyticsClient = nullClient{}
+		return
+	}
+
+	var err error
+	analyticsClient, err = analytics.NewClient(analytics.Config{
+		WriteKey:        segmentKey,
+		SegmentEndpoint: segmentEndpoint,
+		App: analytics.AppInfo{
+			Name:      "akita-cli",
+			Version:   version.ReleaseVersion().String(),
+			Build:     version.GitVersion(),
+			Namespace: "",
+		},
+		IsLoggingEnabled: false,
+		// IsMixpanelEnabled: false,  -- irrelevant for us, leaving at default value
+		BatchSize: 1, // disable batching
+	})
+	if err != nil {
+		printer.Stderr.Errorf("Telemetry unavailable; error setting up Segment client: %v\n", err)
+		analyticsClient = nullClient{}
+	}
+}
+
+func getDistinctID() string {
+	// If we have a user email, use that!
+	// Otherwise use the configured API Key.
+	// Failing that, try to the user name and host name?
+
+	id := os.Getenv("AKITA_SEGMENT_DISTINCT_ID")
+	if id != "" {
+		return id
+	}
+
+	// TODO: every command creates its client ID separately. That makes it
+	// impossible to use the same one here without major refactoring.
+	// So we use the all-zeros ID so it's obviously suspect.
+	frontClient := rest.NewFrontClient(rest.Domain, GetClientID())
+	userResponse, err := frontClient.GetUser(context.Background())
```

Comment:
Should we use a timeout here?

Reply:
Fixed in https://github.com/akitasoftware/akita-cli/pull/155/commits/15edda28ca047f11bdd13abe447c1b651785fb49